### ISSUE TITLE
panic handling request textDocument/completion: Cannot create token from reparsed node of kind KindHeritageClause

### ISSUE THREAD

**Issue by gabritto** [microsoft/typescript-go#3461]
Copied from `@typescript-bot` results on PR #3454.

Source PR: https://github.com/microsoft/typescript-go/pull/3454

This error was reported in 1 bot report section(s) across 1 affected repo(s).

Source comments:
- https://github.com/microsoft/typescript-go/pull/3454#issuecomment-4285342822

Affected repos:
- siyuan-note/siyuan

Representative raw report:

### panic handling request textDocument/completion: Cannot create token from reparsed node of kind KindHeritageClause

Source comment: https://github.com/microsoft/typescript-go/pull/3454#issuecomment-4285342822

```
textDocument/completion
github.com/microsoft/typescript-go/internal/lsp.(*Server).recover(0x99ac686d508, 0x99aef6836b0)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:940 +0x47
github.com/microsoft/typescript-go/internal/ast.(*SourceFile).GetOrCreateToken(0x99aefbde008, 0x4f, 0x2b2, 0x2b7, 0x99aef1c2fc0, 0x0)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/ast/ast.go:2739 +0x2fa
github.com/microsoft/typescript-go/internal/astnav.findRightmostValidToken.func1(0x99aef1c2fc0, 0x2b7)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:563 +0x6fd
github.com/microsoft/typescript-go/internal/astnav.findRightmostValidToken(0x99aef1c2fc0?, 0x99aefbde008?, 0x99ad8982d58?, 0x5ee7a5?, 0x30?)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:590 +0x68
github.com/microsoft/typescript-go/internal/astnav.FindPrecedingTokenEx.func1(0x99aef1c2fc0)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:417 +0x245
github.com/microsoft/typescript-go/internal/astnav.FindPrecedingTokenEx(0x99aefbda2d8?, 0x99aefbde008?, 0x0?, 0x0?)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:429 +0x63
github.com/microsoft/typescript-go/internal/astnav.getTokenAtPosition(0x99aefbde008, 0x2db, 0x0, 0x1131e68)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:176 +0x2ef
github.com/microsoft/typescript-go/internal/astnav.GetTouchingPropertyName(...)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/astnav/tokens.go:12
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).getCompletionData(_, {_, _}, _, _, _, {{{0x0, 0x4, 0x4, {0x1132e48, ...}, ...}, ...}, ...}, ...)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/ls/completions.go:522 +0x27d
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).getCompletionsAtPosition(0x99aed9c9a40, {0x1143338, 0x99aef683800}, 0x99aefbde008, 0x2db, 0x0)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/ls/completions.go:347 +0x4ea
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).ProvideCompletion(0x99aed9c9a40, {0x1143338, 0x99aef683740}, {0x99aebf7ba40?, 0x0?}, {0xf06fb600?, 0x99a?}, 0x99aef447970)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/ls/completions.go:47 +0x216
github.com/microsoft/typescript-go/internal/lsp.(*Server).handleCompletion(0x99acf3fd868?, {0x1143338?, 0x99aef683740?}, 0x99aebe03c20?, 0xe47f9ae879ce0432?)
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:1474 +0x39
github.com/microsoft/typescript-go/internal/lsp.init.func1.registerLanguageServiceWithAutoImportsRequestHandler[...].33.1.1()
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:836 +0xf3
github.com/microsoft/typescript-go/internal/project.(*Session).WithLanguageServiceAndSnapshot.func1()
	/mnt/vss/_work/1/s/typescript-go-3454/internal/project/session.go:1015 +0x5d
github.com/microsoft/typescript-go/internal/lsp.(*Server).handleRequestOrNotification.func1()
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:664 +0x65
github.com/microsoft/typescript-go/internal/lsp.(*Server).dispatchLoop.func3()
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:532 +0x2c
created by github.com/microsoft/typescript-go/internal/lsp.(*Server).dispatchLoop in goroutine 10
	/mnt/vss/_work/1/s/typescript-go-3454/internal/lsp/server.go:531 +0x531
```

<h4>Affected repos</h4>
<details>
<summary><a href="https://github.com/siyuan-note/siyuan">siyuan-note/siyuan</a></summary>
Raw error text: <code>RepoResults7/siyuan-note.siyuan.rawError.txt</code> in the <a href="https://typescript.visualstudio.com/TypeScript/_build/results?buildId=167767&view=artifacts&type=publishedArtifacts">artifact folder</a> <br />
Replay commands: <code>RepoResults7/siyuan-note.siyuan.replay.txt</code> in the <a href="https://typescript.visualstudio.com/TypeScript/_build/results?buildId=167767&view=artifacts&type=publishedArtifacts">artifact folder</a>
<h4>Old server result</h4>

```
textDocument/completion
github.com/microsoft/typescript-go/internal/lsp.(*Server).recover(0x2751cf04b508, 0x2751e5122ed0)
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:940 +0x47
github.com/microsoft/typescript-go/internal/ast.(*SourceFile).GetOrCreateToken(0x2751e41e3888, 0x4f, 0x2b2, 0x2b7, 0x2751e3957040, 0x0)
	/mnt/vss/_work/1/s/typescript-go-main/internal/ast/ast.go:2739 +0x2fa
github.com/microsoft/typescript-go/internal/astnav.findRightmostValidToken.func1(0x2751e3957040, 0x2b7)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:563 +0x6fd
github.com/microsoft/typescript-go/internal/astnav.findRightmostValidToken(0x2751e3957040?, 0x2751e41e3888?, 0x2751dfd76d58?, 0x5ee7a5?, 0x30?)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:590 +0x68
github.com/microsoft/typescript-go/internal/astnav.FindPrecedingTokenEx.func1(0x2751e3957040)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:417 +0x245
github.com/microsoft/typescript-go/internal/astnav.FindPrecedingTokenEx(0x2751e39839e8?, 0x2751e41e3888?, 0x0?, 0x0?)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:429 +0x63
github.com/microsoft/typescript-go/internal/astnav.getTokenAtPosition(0x2751e41e3888, 0x2db, 0x0, 0x1131ca8)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:176 +0x2ef
github.com/microsoft/typescript-go/internal/astnav.GetTouchingPropertyName(...)
	/mnt/vss/_work/1/s/typescript-go-main/internal/astnav/tokens.go:12
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).getCompletionData(_, {_, _}, _, _, _, {{{0x0, 0x4, 0x4, {0x1132c88, ...}, ...}, ...}, ...}, ...)
	/mnt/vss/_work/1/s/typescript-go-main/internal/ls/completions.go:522 +0x27d
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).getCompletionsAtPosition(0x2751e35f7c00, {0x1143178, 0x2751e4b6ca80}, 0x2751e41e3888, 0x2db, 0x0)
	/mnt/vss/_work/1/s/typescript-go-main/internal/ls/completions.go:347 +0x4ea
github.com/microsoft/typescript-go/internal/ls.(*LanguageService).ProvideCompletion(0x2751e35f7c00, {0x1143178, 0x2751e4b6c9c0}, {0x2751e3680ff0?, 0x2751d86e65c0?}, {0xaa17d7?, 0x0?}, 0x2751e511c340)
	/mnt/vss/_work/1/s/typescript-go-main/internal/ls/completions.go:47 +0x216
github.com/microsoft/typescript-go/internal/lsp.(*Server).handleCompletion(0xcfb6263f?, {0x1143178?, 0x2751e4b6c9c0?}, 0x2751d86e65b0?, 0x466f0e?)
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:1474 +0x39
github.com/microsoft/typescript-go/internal/lsp.init.func1.registerLanguageServiceWithAutoImportsRequestHandler[...].33.1.1()
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:836 +0xf3
github.com/microsoft/typescript-go/internal/project.(*Session).WithLanguageServiceAndSnapshot.func1()
	/mnt/vss/_work/1/s/typescript-go-main/internal/project/session.go:996 +0x5d
github.com/microsoft/typescript-go/internal/lsp.(*Server).handleRequestOrNotification.func1()
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:664 +0x65
github.com/microsoft/typescript-go/internal/lsp.(*Server).dispatchLoop.func3()
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:532 +0x2c
created by github.com/microsoft/typescript-go/internal/lsp.(*Server).dispatchLoop in goroutine 3
	/mnt/vss/_work/1/s/typescript-go-main/internal/lsp/server.go:531 +0x531
```
<h4>Last few requests</h4>

```json
{"kind":"request","method":"textDocument/prepareCallHierarchy","params":{"textDocument":{"uri":"@PROJECT_ROOT_URI@/app/src/asset/pdf/genericl10n.js"},"position":{"line":22,"character":3}}}
{"kind":"request","method":"textDocument/completion","params":{"textDocument":{"uri":"@PROJECT_ROOT_URI@/app/src/asset/pdf/genericl10n.js"},"position":{"line":22,"character":3},"context":{"triggerKind":1}}}
{"kind":"request","method":"completionItem/resolve","params":{"label":"@abstract","kind":14,"sortText":"11","commitCharacters":[".",",",";"],"data":{"fileName":"@PROJECT_ROOT@/app/src/asset/pdf/genericl10n.js","position":677,"name":"@abstract"}}}
{"kind":"request","method":"textDocument/completion","params":{"textDocument":{"uri":"@PROJECT_ROOT_URI@/app/src/asset/pdf/genericl10n.js"},"position":{"line":24,"character":30},"context":{"triggerKind":1}}}

```

<h4>Repro steps</h4>

```bash
#!/bin/bash

git clone https://github.com/siyuan-note/siyuan --recurse-submodules
git -C "./siyuan" reset --hard 62eed37a3263e3816354beea4041d82419b92b9d
downloadUrl=$(curl -s "https://typescript.visualstudio.com/TypeScript/_apis/build/builds/167767/artifacts?artifactName=RepoResults7&api-version=7.0" | jq -r ".resource.downloadUrl")
wget -O RepoResults7.zip "$downloadUrl"
unzip -p RepoResults7.zip RepoResults7/siyuan-note.siyuan.replay.txt > siyuan-note.siyuan.replay.txt
npm install --no-save @typescript/server-replay
```

To run the repro:
```bash
# `npx tsreplay --help` to learn about helpful switches for debugging, logging, etc.
npx tsreplay ./siyuan ./siyuan-note.siyuan.replay.txt <PATH_TO_tsserver.js>
```

</details>

Additional raw reports are available in the source comment links above.
